aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,971 @@
+/// Type of large-scale search trees.
+///
+/// It internally uses BTree to organize the search tree data structure for keys. Comparing with
+/// other common search trees like AVL or Red-black tree, a BTree node has more children, and packs
+/// more metadata into one node, which is more disk friendly (and gas friendly).
+
+module aptos_std::smart_tree {
+    use aptos_std::bucket_table::{Self, BucketTable};
+    use std::option::{Self, Option};
+    use std::vector;
+
+    // Internal errors.
+    const E_INTERNAL: u64 = 0;
+    // The tree is not empty, and cannot be destroyed.
+    const E_TREE_NOT_EMPTY: u64 = 1;
+    // The tree is too big for insertion.
+    const E_TREE_TOO_BIG: u64 = 2;
+    // The provided parameter is invalid.
+    const E_INVALID_PARAMETER: u64 = 3;
+
+    const NULL_INDEX: u64 = 0;
+    const DEFAULT_ORDER : u8 = 32;
+    const MAX_NUM_NODES : u64 = 1048576;
+
+    /// A node of the SmartTree.
+    struct Node has drop, store {
+        // Whether this node is a leaf node.
+        is_leaf: bool,
+        // The node index of its parent node, or NULL_INDEX if it doesn't have parent.
+        parent: u64,
+        // The children of the nodes. (When the node is leaf node, all keys of the node is stored in children.max_key)
+        children: vector<Child>,
+        // The node index of its previous node at the same layer, or NULL_INDEX if it doesn't have a previous node.
+        prev: u64,
+        // The node index of its next node at the same layer, or NULL_INDEX if it doesn't have a previous node.
+        next: u64,
+    }
+
+    /// The metadata of a child of a node.
+    struct Child has copy, drop, store {
```

Comment:
I originally wrote it in that way to have two vectors, then I found it's too verbose since for most of the time node_index and max_key need to be maintained together.